### PR TITLE
sock: change "no timeout" value from 0 to UINT32_MAX

### PR DESCRIPTION
**Pull request by kaspar030**
IMHO specifying a timeout of "0" for "don't wait at all" makes much more sense.

### CODE REVIEW

**Comment by OlegHahm** [comment]
I must confess to know next to nothing about `sock`, but this change makes sense to me.

**Comment by miri64** [comment]
What does 0 then do?

**Comment by OlegHahm** [comment]
> IMHO specifying a timeout of "0" for "don't wait at all" makes much more sense.

**Comment by OlegHahm** [comment]
I.e. return immediately - in any case.

**Comment by miri64** [comment]
>> IMHO specifying a timeout of "0" for "don't wait at all" makes much more sense.

I should read more...

Nevertheless, I still don't think it's a good idea. `0` means to me no timeout at all, not 0 sec timeout. Also, though my assembly is quite rusty, I think `if (t == 0)` is translated into fewer instructions than `if (t < UINT_MAX32)`.

**Comment by miri64** [comment]
> I.e. return immediately - in any case.

I wouldn't know how to implement such a case with stacks where `0` means no timeout... I would set it to one, maybe...

File: sys/include/net/sock/ip.h [hunk 1]
```diff
@@ -389,10 +390,9 @@ int sock_ip_get_remote(sock_ip_t *sock, sock_ip_ep_t *ep);
  * @param[out] data     Pointer where the received data should be stored.
  * @param[in] max_len   Maximum space available at @p data.
  * @param[in] timeout   Timeout for receive in microseconds.
- *                      This value can be ignored (no timeout) if the
- *                      @ref sys_xtimer module is not present or the
- *                      implementation does not support timeouts on its own.
- *                      May be 0 for no timeout.
+ *                      If 0 and no data is available, return immediately.
```
Comment (miri64) [inline sys/include/net/sock/ip.h]: You mean "the operation is non-blocking"? "return immediately" sounds more like implementation-speak...

**Comment by kaspar030** [comment]
> 0 means to me no timeout at all, not 0 sec timeout. 

If you read the parameter as "how much time to wait at most until data arrives", "0" meaning "don't wait at all" is quite natural, don't you think?

> Also, though my assembly is quite rusty, I think if (t == 0) is translated into fewer instructions than if (t < UINT_MAX32).

You might save one instruction...

File: sys/include/net/sock/ip.h (hunk 1 above)
Comment (kaspar030) [inline sys/include/net/sock/ip.h]: > I wouldn't know how to implement such a case with stacks where 0 means no timeout... I would set it to one, maybe...

Just add ```timeout += 1;``` at a strategically useful location. ;)

**Comment by kaspar030** [comment]
> I wouldn't know how to implement such a case with stacks where 0 means no timeout... I would set it to one, maybe...

Just add ```timeout += 1;``` at a strategically useful location. ;)

File: sys/include/net/sock/ip.h (hunk 1 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: > Yes, better

Then this is null and void:

> I wouldn't know how to implement such a case with stacks where 0 means no timeout... I would set it to one, maybe...

Then please add the return value `-EWOULDBLOCK` or `-EAGAIN`.

> I'll change if you approve the change in general.

I still see some implementation confusions, but I guess that is the reviewers job to spot these ;-)

**Comment by miri64** [comment]
Marginally related: do you remember, why we decided against a configurable timeout for `sock_accept()`? It's basically the receive function of a `sock_tcp_queue_t`.

**Comment by kaspar030** [comment]
> Marginally related: do you remember, why we decided against a configurable timeout for sock_accept()? It's basically the receive function of a sock_tcp_queue_t.

Actually, no, don't even remember discussing it. Seems to be missing.

**Comment by miri64** [comment]
I have a slither of a memory, that Simon brought it up...

**Comment by kaspar030** [comment]
> You mean "the operation is non-blocking"? "return immediately" sounds more like implementation-speak...

Actually "the operation is non-blocking" seems less clear to me than "returns immediately". Would that extra s for return change your mind, too?

**Comment by miri64** [comment]
Yes.

**Comment by kaspar030** [comment]
@miri64 I think I got all your comments?

File: sys/include/net/sock/ip.h [hunk 2]
```diff
@@ -389,10 +390,9 @@ int sock_ip_get_remote(sock_ip_t *sock, sock_ip_ep_t *ep);
  * @param[out] data     Pointer where the received data should be stored.
  * @param[in] max_len   Maximum space available at @p data.
  * @param[in] timeout   Timeout for receive in microseconds.
- *                      This value can be ignored (no timeout) if the
- *                      @ref sys_xtimer module is not present or the
- *                      implementation does not support timeouts on its own.
- *                      May be 0 for no timeout.
+ *                      If 0 and no data is available, returns immediately.
```
Comment (miri64) [inline sys/include/net/sock/ip.h]: Okay, now I'm seeing this I immediately wonder: "What returns immediately" (the half-sentence is missing a subject ;-)).

File: sys/include/net/sock/tcp.h [hunk 3]
```diff
@@ -191,10 +191,9 @@ int sock_tcp_accept(sock_tcp_queue_t *queue, sock_tcp_t **sock);
  *                      truncated and the remaining data can be retrieved
  *                      later on.
  * @param[in] timeout   Timeout for receive in microseconds.
- *                      This value can be ignored (no timeout) if the
- *                      @ref sys_xtimer module is not present and the
- *                      implementation does not support timeouts on its own.
- *                      May be 0 for no timeout.
+ *                      If 0 and no data is available, returns immediately.
```
Comment (miri64) [inline sys/include/net/sock/tcp.h]: Dito

File: sys/include/net/sock/udp.h [hunk 4]
```diff
@@ -377,10 +377,9 @@ int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep);
  * @param[out] data     Pointer where the received data should be stored.
  * @param[in] max_len   Maximum space available at @p data.
  * @param[in] timeout   Timeout for receive in microseconds.
- *                      This value can be ignored (no timeout) if the
- *                      @ref sys_xtimer module is not present or the
- *                      implementation does not support timeouts on its own.
- *                      May be 0 for no timeout.
+ *                      If 0 and no data is available, returns immediately.
```
Comment (miri64) [inline sys/include/net/sock/udp.h]: Dito

File: sys/include/net/sock/udp.h [hunk 5]
```diff
@@ -94,11 +94,11 @@
  *     }
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
- * The application then waits indefinitely for an incoming message in
- * `buf` from `remote`. If we want to timeout this wait period we could
- * alternatively set the `timeout` parameter of @ref sock_udp_recv() to a
- * value `> 0`. If an error occurs on receive we just ignore it and continue
- * looping.
+ * The application then waits indefinitely for an incoming message in `buf`
```
Comment (miri64) [inline sys/include/net/sock/udp.h]: Minor question: Why did you change the line wrapping here?

File: sys/include/net/sock/udp.h [hunk 6]
```diff
@@ -94,11 +94,11 @@
...
+ * from `remote`. If we want to timeout this wait period we could alternatively
+ * set the `timeout` parameter of @ref sock_udp_recv() to a value != @ref
```
Comment (miri64) [inline sys/include/net/sock/udp.h]: I prefer the `@ref` and `SOCK_NO_TIMEOUT` to be on the same line

File: sys/include/net/sock/ip.h (hunk 2 above)
Comment (kaspar030) [inline sys/include/net/sock/ip.h]: so "the function returns immediately." ?

File: sys/include/net/sock/udp.h (hunk 5 above)
Comment (kaspar030) [inline sys/include/net/sock/udp.h]: I usually leave the line wrapping to vim. Mark the paragraph, press "gq", good.

File: sys/include/net/sock/ip.h (hunk 2 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: :+1:

File: sys/include/net/sock/ip.h [hunk 7]
```diff
@@ -407,6 +407,7 @@ int sock_ip_get_remote(sock_ip_t *sock, sock_ip_ep_t *ep);
  * @return  -EPROTO, if source address of received packet did not equal
  *          the remote of @p sock.
  * @return  -ETIMEDOUT, if @p timeout expired.
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
```
Comment (miri64) [inline sys/include/net/sock/ip.h]: Nitpick-alert: The other return-values are lexicographically ordered ;-)

File: sys/include/net/sock/tcp.h [hunk 8]
```diff
@@ -207,6 +206,7 @@ int sock_tcp_accept(sock_tcp_queue_t *queue, sock_tcp_t **sock);
  *          point of @p sock.
  * @return  -ENOTCONN, when @p sock is not connected to a remote end point.
  * @return  -ETIMEDOUT, if @p timeout expired.
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
```
Comment (miri64) [inline sys/include/net/sock/tcp.h]: Nitpick-alert: The other return-values are lexicographically ordered ;-)

File: sys/include/net/sock/udp.h [hunk 9]
```diff
@@ -395,6 +394,7 @@ int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep);
  * @return  -EPROTO, if source address of received packet did not equal
  *          the remote of @p sock.
  * @return  -ETIMEDOUT, if @p timeout expired.
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
```
Comment (miri64) [inline sys/include/net/sock/udp.h]: Nitpick-alert: The other return-values are lexicographically ordered ;-)

Comment (kaspar030) [inline sys/include/net/sock/udp.h]: :)

File: sys/include/net/sock/udp.h (hunk 6 above)
Comment (kaspar030) [inline sys/include/net/sock/udp.h]: Yeah, but together with the line change above the line would cross the 80 characters mark...

Comment (miri64) [inline sys/include/net/sock/udp.h]: You could put the line change before the `@ref` or even the `value !=` ;-)

File: sys/include/net/sock/ip.h [hunk 10]
```diff
@@ -97,8 +97,8 @@
  * The application then waits indefinitely for an incoming message in
  * `buf` from `remote`. If we want to timeout this wait period we could
  * alternatively set the `timeout` parameter of @ref sock_ip_recv() to a
- * value `> 0`. If an error occurs on receive we just ignore it and continue
- * looping.
+ * value `!= SOCK_NO_TIMEOUT`. If an error occurs on receive we just ignore it
+ * and continue looping.
```
Comment (miri64) [inline sys/include/net/sock/ip.h]: Wait... the code above needs adaption to to be in line with that text.

File: sys/include/net/sock/udp.h (hunk 5 above)
Comment (miri64) [inline sys/include/net/sock/udp.h]: Here too.

**Comment by miri64** [comment]
I hope I found everything now ^^"

**Comment by kaspar030** [comment]
> I hope I found everything now ^^"

Could we leave the formatting to "fill lines up to 80 chars, wrap before word"?

**Comment by miri64** [comment]
Yapp

File: sys/include/net/sock/ip.h (hunk 10 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: In the snippet directly above too, please.

File: sys/include/net/sock/udp.h (hunk 5 above)
Comment (miri64) [inline sys/include/net/sock/udp.h]: .

**Comment by kaspar030** [comment]
I don't see what you mean. Could you point me to file and line?

**Comment by miri64** [comment]
> I don't see what you mean. Could you point me to file and line?

Oops, sorry. My bad... You already changed it but I was confused due to the diff layout.
